apps/backend/src/action/logic/remove-path-card.ts

```diff
@@ -1,27 +1,108 @@
-import { ResultAsync, errAsync } from "neverthrow";
+import { ResultAsync, ok, err } from "neverthrow";
 import { EventSource } from "~/models/event";
 import { RockfallCommand } from "~/action/command";
-import { PathCardHasBeenRemovedEvent, Event } from "~/action/event";
+import { PathCardHasBeenRemovedEvent } from "~/action/event";
+import type { BoardEvent } from "~/board/event";
+import type { RepositoryReadError } from "~/board/logic/get-current-placements";
+import getCurrentPlacements from "~/board/logic/get-current-placements";
+import { Placement } from "~/models/placement";
+import { flow, pipe } from "fp-ts/lib/function";
+import { prop, error, always } from "~/utils";
+import { PathCard } from "~/models/card";
+import * as Vec from "~/models/vec";
+import * as Array from "fp-ts/Array";
+import * as Either from "fp-ts/Either";
+
+export type BoardCardEvent = BoardEvent | PathCardHasBeenRemovedEvent;
+
+const TargetCannotBeRemovedError = error("TargetCannotBeRemovedError");
+const RepositoryWriteError = error("RepositoryWriteError");
+export type RepositoryWriteError = ReturnType<typeof RepositoryWriteError>;
+export type TargetCannotBeRemovedError = ReturnType<
+  typeof TargetCannotBeRemovedError
+>;
+export type RemovePathCardError =
+  | TargetCannotBeRemovedError
+  | RepositoryWriteError
+  | RepositoryReadError;
 
 export interface RemovePathCard {
-  (source: EventSource<Event>, command: RockfallCommand): ResultAsync<
+  (source: EventSource<BoardCardEvent>, command: RockfallCommand): ResultAsync<
     PathCardHasBeenRemovedEvent,
-    Error
+    RemovePathCardError
   >;
 }
 
+const NonRemovablePathCard = Object.freeze([
+  PathCard.START,
+  PathCard.GOAL_COAL_BOTTOM_LEFT,
+  PathCard.GOAL_COAL_BOTTOM_RIGHT,
+  PathCard.GOAL_GOLD,
+]);
+
+const isRemovablePathCard = Either.fromPredicate<
+  Placement,
+  TargetCannotBeRemovedError
+>(
+  (placement) => !NonRemovablePathCard.includes(placement.card),
+  (placement) =>
+    TargetCannotBeRemovedError(
+      `the path card ${placement.card} cannot be removed`
+    )
+);
+
+const findPlacementByPosition =
+  (command: RockfallCommand) => (board: Placement[]) =>
+    Array.findFirst<Placement>(
+      flow(prop("position"), Vec.eq(command.data.position))
+    )(board);
+
+const checkPathCardHasBeenPlaced =
+  (command: RockfallCommand) => (board: Placement[]) =>
+    pipe(
+      board,
+      findPlacementByPosition(command),
+      Either.fromOption(() =>
+        TargetCannotBeRemovedError("empty placement cannot be removed")
+      ),
+      Either.chain(isRemovablePathCard),
+      Either.matchW(err, always(ok(command)))
+    );
+
+const appendEventToEventSource =
```

Comment:
@JohnsonMao 
To improve the code readability.
Can you help me update it by the following code? 
```ts
    pipe(
      PathCardHasBeenRemovedEvent(command.data),
      (event) =>
        ResultAsync.fromPromise(
          source.append(event),
          always(RepositoryWriteError("failed to write event to repository"))
        )
        .map(() => event)
    );
```